Source/DafnyLanguageServer/Workspace/Compilation.cs

```diff
@@ -0,0 +1,461 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reactive;
+using System.Reactive.Disposables;
+using System.Reactive.Subjects;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Boogie;
+using Microsoft.Dafny.LanguageServer.Language;
+using Microsoft.Dafny.LanguageServer.Util;
+using Microsoft.Dafny.LanguageServer.Workspace.Notifications;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.Dafny.LanguageServer.Workspace;
+
+public delegate Compilation CreateCompilation(
+  DafnyOptions options,
+  ExecutionEngine boogieEngine,
+  CompilationInput compilation);
+
+/// <summary>
+/// The compilation of a single document version.
+/// The document will be parsed, resolved, translated to Boogie and verified.
+///
+/// Compilation may be configured to pause after translation,
+/// requiring a call to Compilation.VerifySymbol for the document to be verified.
+///
+/// Compilation is agnostic to document updates, it does not handle the migration of old document state.
+/// </summary>
+public class Compilation : IDisposable {
+
+  private readonly ILogger logger;
+  private readonly ITextDocumentLoader documentLoader;
+  private readonly IProgramVerifier verifier;
+
+  private readonly TaskCompletionSource started = new();
+  private readonly CancellationTokenSource cancellationSource;
+  private readonly ConcurrentDictionary<Uri, ConcurrentStack<DafnyDiagnostic>> staticDiagnostics = new();
+  public DafnyDiagnostic[] GetDiagnosticsForUri(Uri uri) =>
+    staticDiagnostics.TryGetValue(uri, out var forUri) ? forUri.ToArray() : Array.Empty<DafnyDiagnostic>();
+
+  /// <summary>
+  /// FilePosition is required because the default module lives in multiple files
+  /// </summary>
+  private readonly LazyConcurrentDictionary<ModuleDefinition,
+    Task<IReadOnlyDictionary<FilePosition, IReadOnlyList<IImplementationTask>>>> translatedModules = new();
+
+  private readonly ConcurrentDictionary<ICanVerify, Unit> verifyingOrVerifiedSymbols = new();
+  private readonly LazyConcurrentDictionary<ICanVerify, Dictionary<string, ImplementationState>> implementationsPerVerifiable = new();
+
+  private TaskCompletionSource verificationCompleted = new();
+  private readonly DafnyOptions options;
+  public CompilationInput Input { get; }
+  private readonly ExecutionEngine boogieEngine;
+
+  private readonly Subject<ICompilationEvent> updates = new();
+  public IObservable<ICompilationEvent> Updates => updates;
+
+  private Program? programAfterParsing;
+  private Program? compiledProgram;
```

Comment:
compiledProgram feels like a program that was emitted to the final target source code. What do you mean by compiledProgram here?

Reply:
Given the context of this class, a Compilation, there is a single AST that is compiled throughout the compilation. This variable contains the AST at its currently most compiled point.

Reply:
I can rename it to `program` but then I feel you might think it's the parsed program, without any changes. I can also rename it to `compilingProgram`, to indicate that's it not finished per se. Or `transformedProgram` or `transformingProgram`.

Renamed to `transformedProgram` for now